Use string scanner with baseparser

naitoh opened the pull request:
Using StringScanner reduces the string copying process and speeds up the process.

And I removed unnecessary methods.

https://github.com/ruby/rexml/actions/runs/7549990000/job/20554906140?pr=105

```
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [x86_64-linux]
Calculating -------------------------------------
                     rexml 3.2.6      master  3.2.6(YJIT)  master(YJIT) 
                 dom       4.868       5.077        8.137         8.303 i/s -     100.000 times in 20.540529s 19.696590s 12.288900s 12.043666s
                 sax      13.597      13.953       19.206        20.948 i/s -     100.000 times in 7.354343s 7.167142s 5.206745s 4.773765s
                pull      15.641      16.918       22.266        25.378 i/s -     100.000 times in 6.393424s 5.910955s 4.491201s 3.940471s
              stream      14.339      15.844       19.810        22.206 i/s -     100.000 times in 6.973856s 6.311350s 5.047957s 4.503244s

Comparison:
                              dom
        master(YJIT):         8.3 i/s 
         3.2.6(YJIT):         8.1 i/s - 1.02x  slower
              master:         5.1 i/s - 1.64x  slower
         rexml 3.2.6:         4.9 i/s - 1.71x  slower

                              sax
        master(YJIT):        20.9 i/s 
         3.2.6(YJIT):        19.2 i/s - 1.09x  slower
              master:        14.0 i/s - 1.50x  slower
         rexml 3.2.6:        13.6 i/s - 1.54x  slower

                             pull
        master(YJIT):        25.4 i/s 
         3.2.6(YJIT):        22.3 i/s - 1.14x  slower
              master:        16.9 i/s - 1.50x  slower
         rexml 3.2.6:        15.6 i/s - 1.62x  slower

                           stream
        master(YJIT):        22.2 i/s 
         3.2.6(YJIT):        19.8 i/s - 1.12x  slower
              master:        15.8 i/s - 1.40x  slower
         rexml 3.2.6:        14.3 i/s - 1.55x  slower
```

- YJIT=ON : 1.02x - 1.14x faster
- YJIT=OFF : 1.02x - 1.10x faster

kou commented on lib/rexml/parseexception.rb:
```diff
@@ -29,7 +29,7 @@ def to_s
         err << "Position: #{position}\n"
         err << "Last 80 unconsumed characters:\n"
-        err << @source.buffer[0..80].force_encoding("ASCII-8BIT").gsub(/\n/, ' ')
+        err << @source.scanner.rest[0..80].force_encoding("ASCII-8BIT").gsub(/\n/, ' ')
```
Hmm... Can we hide `StringScanner` as a `Source`'s implementation details?

naitoh replied:
OK
I See.

naitoh replied:
@kou 
I hid the `StringScanner` by removing the `attr_reader :buffer` and adding a similar `def buffer` and `def buffer=` interface.

kou commented on lib/rexml/parsers/baseparser.rb:
```diff
@@ -259,7 +259,7 @@ def pull_event
             @document_status = :after_doctype
             if @source.encoding == "UTF-8"
-              @source.buffer.force_encoding(::Encoding::UTF_8)
+              @source.buffer = @source.buffer.force_encoding(::Encoding::UTF_8)
```
How about providing `Source#buffer_encoding=` instead of providing `Source#buffer=`?

naitoh replied:
Instead of providing `Source#buffer=`, I would like to provide `Source#force_encoding`.

kou replied:
I don't want to use `force_encoding` here because it doesn't change `REXML::Source#encoding` that is defined in `REXML::Encoding`. But `String#force_encoding` changes `String#encoding`. They will confuse us.

naitoh replied:
I see.
Change to `Source#buffer_encoding=`.

kou commented on lib/rexml/source.rb:
```diff
@@ -181,65 +151,40 @@ def initialize(arg, block_size=500, encoding=nil)
     def read
       begin
-        @buffer << readline
+        @scanner.string = @scanner.rest + readline
```
Can we use `@scanner << readline` here?

naitoh replied:
Changing to `@scanner << readline` causes the following error in JRuby.

https://github.com/ruby/rexml/actions/runs/7434514894/job/20228750659#step:4:43
```
Error: test_rexml(REXMLTests::TestIssuezillaParsing):
  REXML::ParseException: No close tag for /issuezilla/issue[118]/activity[2]
  Line: -1
  Position: -1
  Last 80 unconsumed characters:
/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/treeparser.rb:28:in `parse'
/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/document.rb:448:in `build'
/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/document.rb:101:in `initialize'
org/jruby/RubyClass.java:904:in `new'
/Users/naitoh/ghq/github.com/naitoh/rexml/test/test_rexml_issuezilla.rb:8:in `block in test_rexml'
      5:     include Helper::Fixture
      6:     def test_rexml
      7:       doc = File.open(fixture_path("ofbiz-issues-full-177.xml")) do |f|
  =>  8:         REXML::Document.new(f)
      9:       end
     10:       ctr = 1
     11:       doc.root.each_element('//issue') do |issue|
org/jruby/RubyIO.java:1179:in `open'
/Users/naitoh/ghq/github.com/naitoh/rexml/test/test_rexml_issuezilla.rb:7:in `test_rexml'
org/jruby/RubyKernel.java:1310:in `catch'
org/jruby/RubyKernel.java:1305:in `catch'
org/jruby/RubyKernel.java:1310:in `catch'
org/jruby/RubyKernel.java:1305:in `catch'
```

I am not sure why the error is occurring, but I am thinking that https://github.com/ruby/strscan/issues/78 may be affected.

kou replied:
It seems that the issue shows that `scanner.string = scanner.rest + XXX` has a problem but `scanner << XXX` doesn't have a problem. So it may not be related...

kou replied:
https://github.com/ruby/strscan/issues/78 has been fixed. Could you try with the latest strscan?

naitoh replied:
I tried, but it did not fix it...

naitoh replied:
@kou 
I found the cause of this problem.

I was getting `Java::JavaLang::OutOfMemoryError: Java heap space` when trying to parse huge [XML](https://github.com/ruby/rexml/blob/master/test/data/ofbiz-issues-full-177.xml) in the DOM.


1. Rewrite `def read` as follows.
```ruby
    def read
      #begin
        @scanner << readline
      #rescue Exception, NameError
      #  @source = nil
      #end
    end
```

2. Rewrite `test/run.rb` as follows.
```ruby
diff --git a/test/run.rb b/test/run.rb
index 089318b..703aeb8 100755
--- a/test/run.rb
+++ b/test/run.rb
@@ -4,7 +4,7 @@ $VERBOSE = true
 
 base_dir = File.dirname(File.expand_path(__dir__))
 lib_dir = File.join(base_dir, "lib")
-test_dir = File.join(base_dir, "test")
+test_dir = "test/test_rexml_issuezilla.rb"
 
 $LOAD_PATH.unshift(lib_dir)
```

3. Run Test
```bash
$ bundle exec rake test
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/bin/jruby test/run.rb
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/lib/ruby/gems/shared/gems/power_assert-2.0.3/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/lib/ruby/gems/shared/gems/power_assert-2.0.3/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
Loaded suite REXMLTests::TestIssuezillaParsing
Started
Finished in 5.459958 seconds.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1 tests, 177 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
0.18 tests/s, 32.42 assertions/s
MacBook-Air:rexml naitoh$ bundle exec rake test
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/bin/jruby test/run.rb
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/lib/ruby/gems/shared/gems/power_assert-2.0.3/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/lib/ruby/gems/shared/gems/power_assert-2.0.3/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
Loaded suite REXMLTests::TestIssuezillaParsing
Started
E
============================================================================================================================================================================================================
Error: test_rexml(REXMLTests::TestIssuezillaParsing): Java::JavaLang::OutOfMemoryError: Java heap space
org.jruby.util.ByteList.ensure(ByteList.java:362)
org.jruby.util.io.EncodingUtils.strBufCat(EncodingUtils.java:1832)
org.jruby.util.io.EncodingUtils.strBufCat(EncodingUtils.java:1817)
org.jruby.util.io.EncodingUtils.encCrStrBufCat(EncodingUtils.java:1940)
org.jruby.util.io.EncodingUtils.encCrStrBufCat(EncodingUtils.java:1867)
org.jruby.RubyString.cat19(RubyString.java:1423)
org.jruby.RubyString.cat19(RubyString.java:1409)
org.jruby.RubyString.append(RubyString.java:2675)
org.jruby.ext.strscan.RubyStringScanner.concat(RubyStringScanner.java:207)
org.jruby.ext.strscan.RubyStringScanner$INVOKER$i$1$0$concat.call(RubyStringScanner$INVOKER$i$1$0$concat.gen)
org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:242)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.source.invokeOther2:\=\^\^(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/source.rb:153)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.source.RUBY$method$read$0(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/source.rb:153)
java.lang.invoke.LambdaForm$DMH/917568725.invokeStatic_L6_L(LambdaForm$DMH)
java.lang.invoke.LambdaForm$MH/1088104996.invokeExact_MT(LambdaForm$MH)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:152)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:148)
org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:212)
org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:193)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.parsers.baseparser.invokeOther188:read(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/baseparser.rb:351)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.parsers.baseparser.RUBY$method$pull_event$0(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/baseparser.rb:351)
java.lang.invoke.LambdaForm$DMH/659059448.invokeStatic_L6_L(LambdaForm$DMH)
java.lang.invoke.LambdaForm$MH/1088104996.invokeExact_MT(LambdaForm$MH)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:152)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:148)
org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:212)
org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:193)
org.jruby.runtime.callsite.CachingCallSite.fcall(CachingCallSite.java:199)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.parsers.baseparser.invokeOther3:pull_event(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/baseparser.rb:183)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.parsers.baseparser.RUBY$method$pull$0(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/baseparser.rb:183)
java.lang.invoke.LambdaForm$DMH/830863979.invokeStatic_L6_L(LambdaForm$DMH)
java.lang.invoke.LambdaForm$MH/1088104996.invokeExact_MT(LambdaForm$MH)
============================================================================================================================================================================================================
Finished in 7.390895 seconds.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1 tests, 0 assertions, 0 failures, 1 errors, 0 pendings, 0 omissions, 0 notifications
0% passed
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
0.14 tests/s, 0.00 assertions/s
rake aborted!
Command failed with status (1): [/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/bin/jruby test/run.rb]
/Users/naitoh/ghq/github.com/naitoh/rexml/Rakefile:9:in `block in <main>'
org/jruby/ext/monitor/Monitor.java:82:in `synchronize'
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/bin/bundle:25:in `<main>'
Tasks: TOP => test
```

kou replied:
OK. Could you add a comment why we use `@scanner.string = @scanner.rest + readline` instead of `@scanner << readline` here?

naitoh replied:
OK.
Added comment.

headius replied:
@naitoh Is there a bug in JRuby I should look at or is this expected given the size of the input?

naitoh replied:
@headius Cc: @kou 
I think the problem is with JRuby's StringScanner, so I have created issue.

https://github.com/ruby/strscan/issues/83

headius replied:
@naitoh Thank you! I will try to fix it today.

headius replied:
ruby/strscan#83 is fixed and will be released soon!

naitoh replied:
@kou
https://github.com/ruby/strscan/pull/84 has been merged into master and I confirmed that JRuby's `@scanner << readline` works.

kou replied:
Thanks!

kou commented on lib/rexml/source.rb:
```diff
@@ -58,98 +65,61 @@ def encoding=(enc)
-        @buffer.force_encoding("ASCII-8BIT")
-        if @buffer[0, 2] == "\xfe\xff"
-          @buffer[0, 2] = ""
+        @scanner.string = @scanner.rest.force_encoding("ASCII-8BIT")
+        if @scanner.rest[0, 2] == "\xfe\xff"
+          @scanner.string = @scanner.rest.delete_prefix("\xfe\xff")
```
Does this work?

```suggestion
        @scanner.string.force_encoding("ASCII-8BIT")
        if @scanner.scan(/\xfe\xff/n)
```

naitoh replied:
It worked, so I fixed it.

kou commented on lib/rexml/source.rb:
```diff
@@ -58,98 +65,61 @@ def encoding=(enc)
           detected_encoding = "UTF-16BE"
-        elsif @buffer[0, 2] == "\xff\xfe"
-          @buffer[0, 2] = ""
+        elsif @scanner.rest[0, 2] == "\xff\xfe"
+          @scanner.string = @scanner.rest.delete_prefix("\xff\xfe")
```
Does this work?

```suggestion
        elsif @scanner.scan(/\xff\xfe/n)
```

naitoh replied:
It worked, so I fixed it.

kou commented on lib/rexml/source.rb:
```diff
@@ -58,98 +65,61 @@ def encoding=(enc)
           detected_encoding = "UTF-16LE"
-        elsif @buffer[0, 3] == "\xef\xbb\xbf"
-          @buffer[0, 3] = ""
+        elsif @scanner.rest[0, 3] == "\xef\xbb\xbf"
+          @scanner.string = @scanner.rest.delete_prefix("\xef\xbb\xbf")
```
Does this work?

```suggestion
        elsif @scanner.scan(/\xef\xbb\xbf/n)
```

naitoh replied:
It worked, so I fixed it.

kou commented on lib/rexml/source.rb:
```diff
@@ -58,98 +65,61 @@ def encoding=(enc)
           detected_encoding = "UTF-16LE"
-        elsif @buffer[0, 3] == "\xef\xbb\xbf"
-          @buffer[0, 3] = ""
+        elsif @scanner.rest[0, 3] == "\xef\xbb\xbf"
+          @scanner.string = @scanner.rest.delete_prefix("\xef\xbb\xbf")
```
Does this work?

```suggestion
        elsif @scanner.scan(/\xef\xbb\xbf/n)
```

naitoh replied:
It worked, so I fixed it.

kou commented on lib/rexml/source.rb:
```diff
@@ -58,98 +65,61 @@ def encoding=(enc)
           detected_encoding = "UTF-16LE"
-        elsif @buffer[0, 3] == "\xef\xbb\xbf"
-          @buffer[0, 3] = ""
+        elsif @scanner.rest[0, 3] == "\xef\xbb\xbf"
+          @scanner.string = @scanner.rest.delete_prefix("\xef\xbb\xbf")
```
Does this work?

```suggestion
        elsif @scanner.scan(/\xef\xbb\xbf/n)
```

naitoh replied:
It worked, so I fixed it.

kou commented on lib/rexml/source.rb:
```diff
@@ -58,98 +65,61 @@ def encoding=(enc)
         end
       ensure
-        @buffer.force_encoding(buffer_encoding)
+        @scanner.string = @scanner.rest.force_encoding(scanner_encoding)
```
Does this work?

```suggestion
        @scanner.string.force_encoding(scanner_encoding)
```

naitoh replied:
It worked, so I fixed it.

kou commented on lib/rexml/source.rb:
```diff
@@ -58,98 +65,61 @@ def encoding=(enc)
       else
         @to_utf = false
-        @buffer.force_encoding ::Encoding::UTF_8
+        @scanner.string = @scanner.rest.force_encoding(::Encoding::UTF_8)
```
Does this work?

```suggestion
        @scanner.string.force_encoding(::Encoding::UTF_8)
```

naitoh replied:
It worked, so I fixed it.

kou commented on lib/rexml/source.rb:
```diff
@@ -181,65 +151,40 @@ def initialize(arg, block_size=500, encoding=nil)
       end
-      rv.taint if RUBY_VERSION < '2.7'
-      rv
+      @scanner.taint if RUBY_VERSION < '2.7'
```
Hmm. Let's remove this line.

naitoh replied:
Removed.

kou commented on lib/rexml/parsers/baseparser.rb:
```diff
@@ -274,8 +274,8 @@ def pull_event
-            match = @source.match( ENTITYDECL, true ).to_a.compact
-            match[0] = :entitydecl
+            match = @source.match( ENTITYDECL, true )
+            match = match.nil? ? [:entitydecl] : [:entitydecl, *match.captures.compact.reject(&:empty?)]
```
Hmm. This assumes that `match` is `StringScanner`.
How about returning `@scanner.captures` instead of `@scanner` by `@source.match`?

naitoh replied:
I believe that https://github.com/ruby/strscan/pull/72 needs to be merged in order to use `@scanner.captures`.

kou replied:
Merged and released.

naitoh replied:
Changed `source#match?` to return `scanner.captures` instead of `@scanner`.

naitoh replied:
Add compact option to `@source.match` with 8bc895562c85e4689544277f5a8168a1822b0b50

Improve processing speed by returning `@scanner.captures.compact` if `@compact=true` and `@scanner` if `compact=false`.

naitoh replied:
Added `match?` and removed `compact` option in 50b305711d1a4401975843a88cd6315ea2d1f9df.

naitoh replied:
Removed `Source#match?` and return `@scanner` in Source#match.

naitoh commented:
@kou 
I have implemented the fix, but for some reason CI fails for head only.
(The error that occurs when strscan 3.0.8 is not used is being output. Do you know why?)

I used `@scanner.captures` instead of `@scanner` in `source#match?`, but the performance was much worse.

I don't think this is a good idea...

https://github.com/ruby/rexml/actions/runs/7510468370/job/20448939679?pr=105

```
                     rexml 3.2.6      master  3.2.6(YJIT)  master(YJIT) 
                 dom       4.981       4.923        7.432         8.192 i/s -     100.000 times in 20.075435s 20.314165s 13.455163s 12.207556s
                 sax      13.060      12.877       18.979        18.779 i/s -     100.000 times in 7.657007s 7.765487s 5.269112s 5.325084s
                pull      16.223      15.909       23.373        22.213 i/s -     100.000 times in 6.164121s 6.285777s 4.278464s 4.501960s
              stream      14.449      14.439       20.160        19.273 i/s -     100.000 times in 6.920919s 6.925817s 4.960348s 5.188534s
```

naitoh commented:
Add compact option to `@source.match` with 8bc895562c85e4689544277f5a8168a1822b0b50

Improve processing speed by returning `@scanner.captures.compact` if `@compact=true` and `@scanner` if `compact=false`.

https://github.com/ruby/rexml/actions/runs/7512802060/job/20453872698?pr=105

```
                     rexml 3.2.6      master  3.2.6(YJIT)  master(YJIT) 
                 dom       5.012       5.100        8.407         8.440 i/s -     100.000 times in 19.951141s 19.607244s 11.894233s 11.848374s
                 sax      13.083      13.761       19.094        21.722 i/s -     100.000 times in 7.643699s 7.266801s 5.237383s 4.603641s
                pull      15.565      16.990       22.291        24.950 i/s -     100.000 times in 6.424717s 5.885687s 4.486175s 4.008003s
              stream      14.348      14.817       19.798        21.514 i/s -     100.000 times in 6.969712s 6.749020s 5.051075s 4.648116s
```

kou commented on rexml.gemspec:
```diff
@@ -55,6 +55,8 @@ Gem::Specification.new do |spec|
 
   spec.required_ruby_version = '>= 2.5.0'
 
+  spec.add_runtime_dependency("strscan", "~> 3.0.8")
```
```suggestion
  spec.add_runtime_dependency("strscan", "=> 3.0.8")
```

kou commented:
> Improve processing speed by returning `@scanner.captures.compact` if `@compact=true` and `@scanner` if `compact=false`.

It seems that calling `@scanner.captures` when `@source.match` result isn't used slows down.

How about providing `Source#match?` that just `true`/`false` instead of matched result/`nil` like `Regexp#match?` and `Regexp#match`?
(And removed the added `compact`.)

We'll use `match?` for many cases, for example:

```diff
diff --git a/lib/rexml/parsers/baseparser.rb b/lib/rexml/parsers/baseparser.rb
index 305b120..610209e 100644
--- a/lib/rexml/parsers/baseparser.rb
+++ b/lib/rexml/parsers/baseparser.rb
@@ -223,13 +223,13 @@ module REXML
             return process_instruction
           when DOCTYPE_START
             base_error_message = "Malformed DOCTYPE"
-            @source.match(DOCTYPE_START, true)
+            @source.match?(DOCTYPE_START, true)
             @nsstack.unshift(curr_ns=Set.new)
             name = parse_name(base_error_message)
-            if @source.match(/\A\s*\[/um, true)
+            if @source.match?(/\A\s*\[/um, true)
               id = [nil, nil, nil]
               @document_status = :in_doctype
-            elsif @source.match(/\A\s*>/um, true)
+            elsif @source.match?(/\A\s*>/um, true)
               id = [nil, nil, nil]
               @document_status = :after_doctype
             else
```

kou commented:
> I have implemented the fix, but for some reason CI fails for head only.
> (The error that occurs when strscan 3.0.8 is not used is being output. Do you know why?)

Hmm. https://github.com/ruby/rexml/pull/105/files#r1451610001 may fix this.

naitoh commented:
> > I have implemented the fix, but for some reason CI fails for head only.
> > (The error that occurs when strscan 3.0.8 is not used is being output. Do you know why?)
> 
> Hmm. https://github.com/ruby/rexml/pull/105/files#r1451610001 may fix this.

It was not fixed...

naitoh commented:
Added `match?` and removed `compact` option in 50b305711d1a4401975843a88cd6315ea2d1f9df. 
Slight improvement over https://github.com/ruby/rexml/pull/105#issuecomment-1890374935 ...

https://github.com/ruby/rexml/actions/runs/7516658356/job/20461969215?pr=105

```
                     rexml 3.2.6      master  3.2.6(YJIT)  master(YJIT) 
                 dom       4.932       4.866        8.354         8.015 i/s -     100.000 times in 20.275441s 20.548772s 11.969668s 12.477243s
                 sax      13.182      13.048       19.279        19.058 i/s -     100.000 times in 7.586344s 7.664254s 5.186958s 5.247065s
                pull      15.635      15.440       22.523        22.527 i/s -     100.000 times in 6.396004s 6.476610s 4.439974s 4.439205s
              stream      14.473      14.247       19.828        19.634 i/s -     100.000 times in 6.909637s 7.019042s 5.043251s 5.093328s
```

kou commented:
OK. It seems that we don't access all captured results in our use case.
Let's remove `Source#match?` and return `@scanner` in `Source#match`.

kou commented:
https://github.com/ruby/ruby/pull/9536 will fix the CI failure.

naitoh commented:
I removed `Source#match?` and return `@scanner` in Source#match.

https://github.com/ruby/rexml/actions/runs/7519306454/job/20467689597?pr=105

```
                      rexml 3.2.6      master  3.2.6(YJIT)  master(YJIT) 
                 dom       4.907       5.070        8.226         8.248 i/s -     100.000 times in 20.377790s 19.723392s 12.157131s 12.123566s
                 sax      12.805      13.697       19.942        20.809 i/s -     100.000 times in 7.809347s 7.300932s 5.014450s 4.805517s
                pull      15.677      16.530       23.046        25.013 i/s -     100.000 times in 6.378722s 6.049606s 4.339091s 3.997863s
              stream      14.310      14.990       20.231        22.022 i/s -     100.000 times in 6.988153s 6.671139s 4.943000s 4.540999s
```


- YJIT=ON : 1.00x - 1.09x faster
- YJIT=OFF : 1.04x - 1.07x faster

kou commented on lib/rexml/parsers/baseparser.rb:
```diff
@@ -274,8 +274,8 @@ def pull_event
-            match = @source.match( ENTITYDECL, true ).to_a.compact
-            match[0] = :entitydecl
+            match = @source.match( ENTITYDECL, true )
+            match = match.nil? ? [:entitydecl] : [:entitydecl, *match.captures.compact]
```
Do we need `match.nil?` check here?
(Is there any case that the above `@source.match()` failed?)

naitoh replied:
If the match.nil? check is removed, all tests succeed.

But, if the string `<!ENTITY>` comes in, `@source.match()` responds with `nil` and `undefined method ``captures' for nil` is raised.

However, since `<!ENTITY>` violates the XML specification and should be treated as an error, can I remove check `match.nil?`?

https://xml.coverpages.org/xmlBNF.html
```
EntityDecl ::= '<!ENTITY' S Name S EntityDef S? '>'   /* General entities */
	| '<!ENTITY' S '%' S Name S EntityDef S? '>'   /* Parameter entities */
```

kou replied:
OK. Could you add a test for the case?

kou replied:
Or we can do it as a separated PR.

naitoh replied:
OK
I added [test](https://github.com/ruby/rexml/compare/995d3e2cc133658e1ec2d57b5f1b5ff2fe0ced72..ba9f7fc08de13ab602cfd698295879853fdba0a8)  for this case.

kou commented on lib/rexml/source.rb:
```diff
@@ -50,6 +49,14 @@ def initialize(arg, encoding=nil)
+      @scanner.rest
+    end
+
+    def buffer_encoding=(str)
+      @scanner.string = str
```
Does this work?

```suggestion
    def buffer_encoding=(encoding)
      @scanner.string.force_encoding(encoding)
```

naitoh replied:
It worked, so I fixed it.

kou commented on lib/rexml/source.rb:
```diff
@@ -58,98 +65,58 @@ def encoding=(enc)
+      else
+        @scanner.check(pattern)
+      end
+      @scanner.matched? ? @scanner : nil
```
Can we avoid `matched?` call by using a return value from `@scanner.scan`/`@scanner.check`?

naitoh replied:
Fixed to not use `matched?`.

kou commented on test/test_core.rb:
```diff
@@ -721,22 +721,6 @@ def test_bad_content
       end
     end
 
-    def test_iso_8859_1_output_function
```
Why do we need remove this test?

naitoh replied:
I was testing `Source#scan`, which is no longer needed, so I removed it as unnecessary.
However, since the purpose seemed to be to test iso-8859-1, it was not appropriate to remove it. Sorry.

Replace `Source#scan` in the test with `Source#match` and the test will work, so modify it in a way that leaves the test.

kou reviewed (commented):
head CI jobs are fixed.

kou commented on test/test_document.rb:
```diff
@@ -173,6 +173,18 @@ def test_empty_value
+          assert_raise(REXML::ParseException) do
+            REXML::Document.new(xml)
+          end
+        end
```
Could you create a new file as `test/parse/test_entity_declaration.rb` and use the same style as https://github.com/ruby/rexml/blob/master/test/parse/test_notation_declaration.rb ?

e.g.:

```ruby
# frozen_string_literal: false
require 'test/unit'
require 'rexml/document'

module REXMLTests
  class TestParseEntityDeclaration < Test::Unit::TestCase
    private
    def xml(internal_subset)
      <<-XML
<!DOCTYPE r SYSTEM "urn:x-henrikmartensson:test" [
#{internal_subset}
]>
<r/>
      XML
    end

    def parse(internal_subset)
      REXML::Document.new(xml(internal_subset)).doctype
    end

    def test_empty
      exception = assert_raise(REXML::ParseException) do
        parse(<<-INTERNAL_SUBSET)
<!ENTITY>
        INTERNAL_SUBSET
       end
        assert_equal(<<-DETAIL.chomp, exception.to_s)
Malformed notation declaration: name is missing
Line: 5
Position: 72
Last 80 unconsumed characters:
 <!ENTITY>  ]> <r/> 
        DETAIL
      end
    end
  end
end
```

naitoh replied:
test/parse/test_entity_declaration.rb was created.

kou commented on lib/rexml/source.rb:
```diff
@@ -181,65 +147,43 @@ def initialize(arg, block_size=500, encoding=nil)
-        @buffer << readline
+        # NOTE: `@scanner << readline` does not free memory, so when parsing huge XML in JRuby's DOM,
+        # out-of-memory error `Java::JavaLang::OutOfMemoryError: Java heap space` occurs.
+        # `@scanner.string = @scanner.rest + readline` frees memory and avoids this problem.
```
```suggestion
        # `@scanner.string = @scanner.rest + readline` frees memory that is already consumed
        # and avoids this problem.
```

kou reviewed (approved):
+1

Could you update the PR description before we merge this?

kou commented:
Thanks!

naitoh commented:
Thanks for your review.
